### ISSUE TITLE
[FEATURE] Field locations and configurability

### ISSUE THREAD

**Issue by johnjaylward** [elastic/ecs-dotnet#77]
**Is your feature request related to a problem? Please describe.**
I was testing out an Alpha release of the NLog Elastic target (https://github.com/markmcdowell/NLog.Targets.ElasticSearch/pull/125) and I found a few items I was curious about.

1. Why is `log.level` a top level field instead of being under the `log` object?
1. I don't see any use of `${callsite}` or `${callsite-linenumber}` in the output. I'd prefer to see those under `log.origin.function` and `log.origin.file.line` respectively assuming they are available at runtime. Looking at the `Elastic.CommonSchema.Log` object definition, this should be doable. Perhaps having a boolean flag on the layout so people could configure to send it or not?
1. As part of the "process" information, I'd like to be able to output the assembly version information so I can track errors that may have happened due to an upgrade. Before I found this PR, I was planning on putting it under the existing `package.version` field since the "process" object didn't seem to have version info. Can that be configured? @snakefoot mentioned that the version number is available under [Service](https://www.elastic.co/guide/en/ecs/current/ecs-service.html). I also saw it available under [Package](https://www.elastic.co/guide/en/ecs/current/ecs-package.html). I think Service makes more sense since the Package fields are probably meant more for tracking installations instead of run logs.
1. ~~For some reason the `log.logger` shows as "Elastic.CommonSchema.NLog" instead of my logger name (i.e. "my.namespace.MyClass"). This seems unexpected. I don't really care what the logging system was that sent the log to Elastic, I care about the system generating the log.~~ fixed in #67 

**Describe the solution you'd like**
Some more configuration options for logging items.
* Flag to log the `CallSite` information under the `Log.Origin` structure
* If there really isn't a good place to store the process version, placing that in a <metadata> tag would be nice.
* Don't place the `log.level` as a top level field and instead use the field on the `log` object.
* Upgrade the documentation for NLog configuration to include configuring `<metadata>` items that can be sent in addition to the standard logs.

**Comment by snakefoot** [comment]
Created #78, so you can do this:

```xml
<layout type="EcsLayout"
    logOriginCallSiteMethod="${callsite}"
    logOriginCallSiteFile="${callsite-filename}"
    logOriginCallSiteLine="${callsite-linenumber}"
 />
```

See also: https://github.com/NLog/NLog/wiki/Callsite-Layout-Renderer

**Comment by Mpdreamz** [comment]
1. The ecs layout renderer makes sure three fields are always first so that if they are used in a different context e.g rendering to disk. They are still somewhat human readable (by having timestamp, level, message) appear first.

**Comment by johnjaylward** [comment]
But JSON object keys are unordered. There is no "first"

**Comment by Mpdreamz** [comment]
That is true and one should never rely on keys being ordered. Here we are in control of the rendering though and we treat this as a feature albeit best effort. This is true for all our ecs logging efforts see e.g

java: https://github.com/elastic/ecs-logging-java#why-ecs-logging

and our common design spec: https://github.com/elastic/ecs-logging/tree/master/spec

**Comment by johnjaylward** [comment]
I see. However, when I view the resulting log in Kibana, the `"log.level"` key shows as a top level item instead of being part of the `log` object. I'm fairly new to Elastic, but I feel like this is breaking the spirit of ECS, which is to have common schema for things. 

If I was trying to grep an output, I think a line that looks like this would be sufficient (assuming no line breaks):
```json
{
    "@timestamp": "2016-05-23T08:05:37.123Z",
    "message": "An error happened!",
    "log": {
        "level": "ERROR",
        "logger": "MyLogger",
        "origin": {
            "file": {
                "name": "App.java",
                "line": 42,
            },
            "function": "methodName"
        }
    },
    "ecs": {"version": "1.4.0"}
}
```

You would just grep for `"log": { "level": "ERROR",` or if you didn't want to depend on key order, grep for `"level": "ERROR"`

When looking in Kibana, the log level would be in the expected schema location.

Am I just thinking too much in the ralational database mindset here and not enough in the "elastic" way of doing things?

**Comment by Mpdreamz** [comment]
2) has been implemented with #78 
3) You should now be able to control the event with #74 

You're chain of thought is not wrong in any way on `log.level`. 

In Elasticsearch searches for `log.level` will work the same whether it came through as `{ "log.level": "", }` or `{ "log": { "level" : "" }`. 

The requirement for it to be separated in the source json is because if you use ECS formatted json to files they will be very hard to read since they are single lined. 

> { "timestamp": "", "log.level": "INFO", "message": "", .... the rest ... }\n
> { "timestamp": "", "log.level": "INFO", "message": "", .... the rest ... }\n
> { "timestamp": "", "log.level": "INFO", "message": "", .... the rest ... }\n
> { "timestamp": "", "log.level": "INFO", "message": "", .... the rest ... }\n
> { "timestamp": "", "log.level": "INFO", "message": "", .... the rest ... }\n
> { "timestamp": "", "log.level": "INFO", "message": "", .... the rest ... }\n
> { "timestamp": "", "log.level": "INFO", "message": "", .... the rest ... }\n
> { "timestamp": "", "log.level": "INFO", "message": "", .... the rest ... }\n

If you use the Logs UI (https://www.elastic.co/guide/en/kibana/master/xpack-logs.html) over Discover in Kibana it will continue to work better over time with ECS logs out of the box. IMO having `log.level` show up separately in Discover is a feature too 😄 

This deviates somewhat from the general guidelines that nested fields should be sent as objects not a map of strings 

e.g

```json
{ 
    "dataset.name": "y",
    "dataset.created" : "date"
}
```

and 

```json
{ 
    "dataset": { "name": "y", "created" : "date" }
}
```

To Elasticsearch are nearly identical when querying the fields, however when retrieving the documents you are obviously presented with different looking documents and these are accessed differently to get to the same value. 

That's why in our mapping of `ECS's` Base object we take extra care to make sure `log.level` is available when you call `b.Log.Level` in .NET. 

So you are right to call it out as an exception which is there for reasons mentioned and the fact its prior art before we started to standardize across multiple ecosystems.

**Comment by johnjaylward** [comment]
Is there a reason you can't do both at the same time? (other than the obvious that it increases the payload)
```json
{
    "@timestamp": "2020...",
    "message": "my log message",
    "log.level": "ERROR",
    "log": {
        "level": "ERROR",
...
}
```
Would that make elastic do weird things?

**Comment by Mpdreamz** [comment]
That could work, mind opening a new issue with that? Would be interesting to get more feedback on that. 

Closing this ticket since most of the initial questions I think are answered. Let me know if you feel I closed this too eagerly!
